include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+				};
+
+				CBasicImageFilterCommon::executePerRegion(commonExecuteData.inImg, decode, commonExecuteData.inRegions.begin(), commonExecuteData.inRegions.end(), clip);
+
+				return true;
+			};
+
+			bool decodeStatus = commonExecute(state, perOutputRegionDecode);
+
+			if (decodeStatus)
+			{
+				const auto arrayLayers = state->inImage->getCreationParameters().arrayLayers;
+				const auto outRegions = state->outImage->getRegions(state->outMipLevel);
+				auto extent = state->outImage->getMipSize(state->outMipLevel);
+				const auto trueOutExtent = asset::VkExtent3D({ extent.x, extent.y, extent.z }); 
+				const auto regionBufferOffset = outRegions.begin()->bufferOffset;
+				auto* imageEntireOutData = reinterpret_cast<uint8_t*>(state->outImage->getBuffer()->getPointer());
+				const size_t singleScratchLayerOffset = (state->extent.width * state->extent.height * state->extent.depth * currentChannelCount); 
+
+				auto sum = [&](core::vectorSIMDu32 readBlockPos) -> void
+				{
+					decltype(readBlockPos) newReadBlockPos = decltype(newReadBlockPos)(readBlockPos.x, trueOutExtent.height - 1 - readBlockPos.y, readBlockPos.z, readBlockPos.w);
+					decodeType* const currentColumnCache = mainColumnCacheScratch + newReadBlockPos.x * currentChannelCount;
+
+					const auto globalIndependentOffset = ((newReadBlockPos.z * state->extent.height + newReadBlockPos.y) * state->extent.width + newReadBlockPos.x);
+					const auto outDataAdressOffsetScratch = globalIndependentOffset * currentChannelCount;
+					const auto layersOffset = singleScratchLayerOffset * newReadBlockPos.w;
+
+					auto finalPixelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(state->mainRowScratchCacheByteSize);
+					auto finalPixel = reinterpret_cast<decodeType*>(finalPixelBuffer->getPointer());
+					memset(finalPixel, 0, finalPixelBuffer->getSize());
+
+					decodeType* decodeBuffer = scratchMemory + outDataAdressOffsetScratch + layersOffset;
+
+					for (uint8_t channel = 0; channel < currentChannelCount; ++channel)
+					{
+						*(currentColumnCache + channel) += *(mainRowCacheScratch + channel) += decodeBuffer[channel];
+						*(finalPixel + channel) = *(currentColumnCache + channel) - (ExclusiveMode ? decodeBuffer[channel] : 0);
+					}
+
+					memcpy(scratchMemory + outDataAdressOffsetScratch + layersOffset, finalPixel, state->mainRowScratchCacheByteSize);
+
+					if (newReadBlockPos.x == trueOutExtent.width - 1) // reset row cache when changing y
+					{
+						if (newReadBlockPos.y == 0) // sum values in (maxX, 0, z) and remove values from column cache
+						{ 
+							memcpy(totalImageCacheScratch + (newReadBlockPos.w * state->extent.depth + newReadBlockPos.z) * state->mainRowScratchCacheByteSize, finalPixel, state->mainRowScratchCacheByteSize);
+							memset(mainColumnCacheScratch, 0, state->mainColumnScratchCacheByteSize);
+						}
+
+						memset(mainRowCacheScratch, 0, state->mainRowScratchCacheByteSize);
+					}
```

Comment:
this is the proper way to do it for a single channel (trivially extendable to 2,3,4), assuming you have z,y,x traversal order as you have right now
```cpp
auto& current = pixel(readBlockPos); // start with current value, x==current_x && y==current_y && z==current_z

current += pixel(readBlockPos-vec4(0,0,1,0)); // add box x<=current_x && y<=current_y && z<current_z
current += pixel(readBlockPos-vec4(0,1,0,0)); // add box x<=current_x && y<current_y && z<=current_z
current -= pixel(readBlockPos-vec4(0,1,1,0)); // remove overlap box x<=current_x && y<current_y && z<current_z
// now I have the sum of all layers below me in the bound x<=current_x && y<=current_y && z<current_z  and the current pixel
// time to add the missing top layer pixels
current += pixel(readBlockPos-vec4(0,1,0,0)); // add box x<current_x && y<=current_y && z<=current_z
current -= pixel(readBlockPos-vec4(0,1,0,1)); // remove overlap box x<current_x && y<=current_y && z<current_z
current -= pixel(readBlockPos-vec4(0,1,1,0)); // remove future overlap box x<current_x && y<current_y && z<=current_z
current += pixel(readBlockPos-vec4(0,1,1,1)); // add box x<current_x && y<current_y && z<current_z
```
where `pixel` is both your decode scratch and your SAT scratch at once (the algorithm is "in place", i.e. doesn't require any extra memory than the input) plus its much faster.

in the above snippet, the `pixel` you touch are actually SAT pixels, because the traversal order guarantees that you've turned these value pixels into SAT pixels in previous iterations

Reply:
`current += pixel(readBlockPos-vec4(0,0,1,0));` it gives you -Z at start and it will crash


I'm wondering if my way of doing it would be good if I limit it to single layer at a time (+ without doing sum across layers, it seems to me the above does that and as far as I remember we weren't to do it)

Reply:
> `current += pixel(readBlockPos-vec4(0,0,1,0));` it gives you -Z at start and it will crash

I noticed I made a typo in the offsets on the last 4 lines.

Everything is fine, but you need to guard the filter for the corner cases properly
```cpp
auto& current = pixel(readBlockPos); // start with current value, x==current_x && y==current_y && z==current_z
const bvec4 axisSafe = greaterThan(readBlockPos,uvec4(0u,0u,0u,0u));
if (axisSafe.z)
    current += pixel(readBlockPos-vec4(0,0,1,0)); // add box x<=current_x && y<=current_y && z<current_z
if (axisSafe.y)
{
    current += pixel(readBlockPos-vec4(0,1,0,0)); // add box x<=current_x && y<current_y && z<=current_z
    if (axisSafe.z)
         current -= pixel(readBlockPos-vec4(0,1,1,0)); // remove overlap box x<=current_x && y<current_y && z<current_z
}
// now I have the sum of all layers below me in the bound x<=current_x && y<=current_y && z<current_z  and the current pixel
// time to add the missing top layer pixels
if (axisSafe.x)
{
    current += pixel(readBlockPos-vec4(1,0,0,0)); // add box x<current_x && y<=current_y && z<=current_z
    if (axisSafe.z)
         current -= pixel(readBlockPos-vec4(1,0,1,0)); // remove overlap box x<current_x && y<=current_y && z<current_z
    if (axisSafe.y)
    {
        current -= pixel(readBlockPos-vec4(1,1,0,0)); // remove future overlap box x<current_x && y<current_y && z<=current_z
        if (axisSafe.z)
            current += pixel(readBlockPos-vec4(1,1,1,0)); // add box x<current_x && y<current_y && z<current_z
     }
}
```

Reply:
corrected